File: 1/2021.03.31/295.Find Median from Data Stream/python/solution.py
class MedianFinder:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.small = []
        self.large = []
        

    def addNum(self, num: int) -> None:
        if len(self.small) <= len(self.large):
            heappush(self.large, num)
            heappush(self.small, -heappop(self.large))
        else:
            heappush(self.small, -num)
            heappush(self.large, -heappop(self.small))
        

    def findMedian(self) -> float:
        if len(self.small) > len(self.large):
            return -self.small[0]
        elif len(self.small) < len(self.large):
            return self.large[0]
        return (self.large[0] - self.small[0]) / 2

File: 1/2021.03.31/295.Find Median from Data Stream/java/solution.java
class MedianFinder {

    private PriorityQueue<Integer> small;
    private PriorityQueue<Integer> large;

    /**
     * initialize your data structure here.
     */
    public MedianFinder() {
        small = new PriorityQueue<>((a, b) -> {
            return b - a;
        });
        large = new PriorityQueue<>();
    }

    public void addNum(int num) {
        if (small.size() >= large.size()) {
            small.offer(num);
            large.offer(small.poll());
        } else {
            large.offer(num);
            small.offer(large.poll());
        }
    }

    public double findMedian() {
        if (small.size() > large.size()) {
            return small.peek();
        } else if (large.size() > small.size()) {
            return large.peek();
        }
        return (small.peek() + large.peek()) / 2.0;
    }
}
